src/qaqc/download_and_flag_fxns/add_biofilm_flag.R

```diff
@@ -21,37 +28,70 @@ add_biofilm_flag <- function(df){
-      #mutate(r_squared4 = data.table::frollapply(mean, n = 288, FUN = progressive_drift, align = "left", fill = NA)) %>%
-      mutate(r_squared5 = data.table::frollapply(mean, n = 288, FUN = progressive_drift, align = "right", fill = NA)) %>%
-      #mutate(r_squared6 = data.table::frollapply(mean, n = 288, FUN = progressive_drift, align = "center", fill = NA)) %>%
-      # r-squared can only be in the positive direction for biofilm drift:
-      add_flag((#r_squared1 >= 0.7 |
-                  r_squared2 >= 0.7 |
-                  #r_squared3 >= 0.7 |
-                #r_squared4 >= 0.8 |
-                  r_squared5 >= 0.8 #|
-                  #r_squared6 >= 0.8
-                )  & !grepl("biofilm", flag), "biofilm") #%>%
-      #select(-c(r_squared1, r_squared2, r_squared3)
-
-    return(drift_tested)
-
-  } else { return(df) }
-  }
+    # Function to check if a selected time window's mean R-squared value is
+    # at least 60% (i.e., fairly high linear relationship with time indicates a "steady" drift)
+    check_too_steady <- function(x) {
+      mean(x) >= 0.60
+    }
 
+    # Function that uses all the functions above to see if a given time window's R-squared with time is strong. If the one-day OR the three-day slope
+    # for a selected parameter is greater than 60%, we determine it has "failed" (i.e., drift seems to exist)
+
+    biofilm_tester <- function(data = sub, col){
+
+      data %>%
+        data.table::data.table() %>%
+        #bind_rows(flagged_data_dfs[["prospect-Turbidity"]]) %>%
+        # dplyr::filter(lubridate::as_date(DT_round) >= "2022-09-10" & lubridate::as_date(DT_round) <= "2022-10-15") %>%
+        dplyr::mutate(r2_s_right = data.table::frollapply(!!sym(col), n = 96, FUN = progressive_drift, align = "right", fill = NA),
+                      r2_s_center = data.table::frollapply(!!sym(col), n = 96, FUN = progressive_drift, align = "left", fill = NA),
+                      r2_l_right = data.table::frollapply(!!sym(col), n = 288, FUN = progressive_drift, align = "right", fill = NA),
+                      r2_l_center = data.table::frollapply(!!sym(col), n = 288, FUN = progressive_drift, align = "left", fill = NA),
+                      tightest_r = pmax(r2_s_center, r2_s_right, r2_l_center, r2_l_right, na.rm = TRUE),
+                      failed = data.table::frollapply(tightest_r, n = 96, FUN = check_too_steady, align = "right", fill = NA)) %>%
+        dplyr::select(!!(col) := "failed",
+                      DT_join)
+    }
 
+    biofouling <- c("Turbidity", "Specific.Conductivity") %>%
+      purrr::map(~biofilm_tester(data = sub, col = .)) %>%
+      dplyr::bind_cols() %>%
+      dplyr::rename(DT_join = DT_join...2) %>%
+      dplyr::mutate(parameter = "Turbidity") %>%
+      dplyr::right_join(., df, by = c("DT_join", "parameter")) %>%
+      # If a steady slope is existing through time for turbidity, but NOT for conductivity, it is likely
+      # sensor bio-fouling
```

Comment:
Suggestion/idea: How long does this function take to run? I don't know how it fits into the workflow right now, and I don't know if I understand the function very well, but I think it might make more sense to have the initial parameter check at the top look for both turbidity and conductivity instead of binding them together before this step. I think that checking if the flags are shared across these parameters could be done in the network_check() function that we have no? 

Though that would mean that we would need to update the network_check() to do intrasensor comparisons and not just intersensor comparisons...

Reply:
I think we want to separate this from our "network_check()" since these comparisons are not happening across the network. Instead they happen at a single site across sensors. But, I like the idea of separating out the stats developed and the flag created from those values!